### PR TITLE
Feature/proof api

### PR DESCRIPTION
**Pull request by minsulee2**
Updated getProof func on #53 .
/get_proof route also checked correctly works.

### CODE REVIEW

File: db/index.js [hunk 1]
```diff
@@ -173,9 +173,9 @@ class DB {
    */
   getProof(dbPath) {
     const fullPath = ChainUtil.parsePath(dbPath);
-    // TODO(minsu): Validate the first label.
-    // TODO(minsu): Implement this.
-    return {};
+    const stateNode = this.getRefForReading(fullPath);
+    if (!!stateNode) return stateNode.getProofHash();
```
Comment (platfowner) [inline db/index.js:177]: In the case we need to convert the value to boolean e.g. `const isNonNull = !!stateNode;`, this `!!` expression would be useful but if we need to check non-null only, `if (stateNode)` would be just enough?

File: db/index.js [hunk 2]
```diff
@@ -173,9 +173,9 @@ class DB {
...
+    else return null;
```
Comment (platfowner) [inline db/index.js:178]: Could you use brace ({}) for if-else statements even though it's simple?

File: test/db.test.js [hunk 3]
```diff
@@ -2030,18 +2039,14 @@ describe("Test Writedatabase with proof", () => {
     rimraf.sync(BLOCKCHAINS_DIR);
   });
 
-  describe("Check proof", () => {
-    it("checks jsObject is correctly set", () => {
-      assert.deepEqual(node.db.getValue("test"), jsObject);
-    });
-
+  describe("Check proof on writeDatabase and setValue", () => {
```
Comment (platfowner) [inline test/db.test.js:2042]: minor: setValue means database write so just 'Check proof for setValue()'?

File: test/db.test.js [hunk 4]
```diff
@@ -2030,18 +2039,14 @@ describe("Test Writedatabase with proof", () => {
...
     it("checks proof hash of /values/test", () => {
       const testNode = node.db.getRefForReading(['values', 'test']);
       const childLabels = testNode.getChildLabels();
       const preimage = `${childLabels[0]}${HASH_DELIMITER}`
           + `${testNode.getChild(childLabels[0]).getProofHash()}`;
       const rootHash = ChainUtil.hashString(ChainUtil.toString(preimage));
```
Comment (platfowner) [inline test/db.test.js:2048]: In db test, you can just use buildProofHashOfStateNode()?

File: test/db.test.js [hunk 5]
```diff
@@ -2051,7 +2056,20 @@ describe("Test Writedatabase with proof", () => {
...
```
Comment (platfowner) [inline test/db.test.js:2058]: ditto

File: test/db.test.js [hunk 6]
```diff
@@ -2021,7 +2021,16 @@ describe("Test Writedatabase with proof", () => {
     node = new Node();
     setDbForTesting(node);
 
-    jsObject = { level0: { level1: { level2: { foo: 'bar', baz: 'caz' } } } };
+    jsObject = {
+      level0: {
+        level1: {
```
Comment (platfowner) [inline test/db.test.js:2026]: Could you add another level1 or level2 nodes? This structure seems to be too simple.

File: test/db.test.js [hunk 7]
```diff
@@ -2051,7 +2056,20 @@ describe("Test Writedatabase with proof", () => {
...
+  describe("getProof", () => {
+    it("tests values, owners, rules and functions", () => {
+      const valuesNode = node.db.getRefForReading(['values', 'test']);
+      const ownersNode = node.db.getRefForReading(['owners', 'test']);
+      const rulesNode = node.db.getRefForReading(['rules', 'test']);
+      const functionNode = node.db.getRefForReading(['functions']);
+      expect(node.db.getProof('/values/test')).to.equal(valuesNode.getProofHash());
+      expect(node.db.getProof('/owners/test')).to.equal(ownersNode.getProofHash());
+      expect(node.db.getProof('/rules/test')).to.equal(rulesNode.getProofHash());
+      expect(node.db.getProof('/functions')).to.equal(functionNode.getProofHash());
```
Comment (platfowner) [inline test/db.test.js:2072]: functions/test?

File: test/db.test.js (hunk 3 above)
Comment (platfowner) [inline test/db.test.js:2042]: Currently we're calling only setValue(), could you call setFunction(), setRule(), setOwner() as well?

File: db/index.js (hunk 1 above)
Comment (minsulee2) [reply db/index.js:177]: Yeah, stateNode can be null is promised here. Updated!

File: db/index.js (hunk 2 above)
Comment (minsulee2) [reply db/index.js:178]: Coding convention checked!

File: test/db.test.js (hunk 6 above)
Comment (minsulee2) [reply test/db.test.js:2026]: Updated!

File: test/db.test.js (hunk 3 above)
Comment (minsulee2) [reply test/db.test.js:2042]: Oh, that's correct!

File: test/db.test.js (hunk 4 above)
Comment (minsulee2) [reply test/db.test.js:2048]: Indeed! it was old one. I've just updated!

File: test/db.test.js (hunk 5 above)
Comment (minsulee2) [reply test/db.test.js:2058]: Updated too!

**Comment by minsulee2** [comment]
PTAL. I've added integration test as well !

**Review by platfowner — APPROVED** [review]
LGTM
with some comments.

File: integration/dev.test.js [hunk 8]
```diff
@@ -236,6 +239,38 @@ describe('API Tests', () => {
...
+        .body.toString('utf-8'));
+      const functionsBody = JSON.parse(
+        syncRequest('GET', server1 + `/get_proof?ref=/${PredefinedDbPaths.FUNCTIONS_ROOT}`)
+        .body.toString('utf-8'));
+      const ownersProof = ownersBody.result;
+      const rulesProof = rulesBody.result;
+      const valuesProof = valuesBody.result;
+      const functionProof = functionsBody.result;
+      // TODO(minsu): it needs to be updated after some updates on db/index.js
+      if (!valuesProof && !ownersProof && !rulesProof && !functionProof) {
```
[the remark this answers is not in the document]
Comment (minsulee2) [reply integration/dev.test.js:264]: I added!

File: integration/dev.test.js [hunk 9]
```diff
@@ -236,6 +239,38 @@ describe('API Tests', () => {
...
+        assert.deepEqual(body, {code: 1, result: null});
+      } else {
+        const preimage = `owners${HASH_DELIMITER}${ownersProof}` + `${HASH_DELIMITER}`
```
[the remark this answers is not in the document]
Comment (minsulee2) [reply integration/dev.test.js]: That's right! updated!

**Comment by minsulee2** [comment]
As discussed, I will investigate proof bugs first and then update it as soon as possible.
Thanks for the review!

**Pull request merged**